### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by ajscimone**

### CODE REVIEW

File: CppTicTacToe/tictactoe.cpp [hunk 1]
```diff
@@ -13,6 +13,7 @@ using namespace std;
 
 // main global variables and initialization follow here
 
+// Test
 int round_counter = 1;
```
Comment (ajscimone) [inline CppTicTacToe/tictactoe.cpp:17]: The first thing I want to say about this code is good advice for C++ and programming in general. Any time there are global variables ( usually just called globals) this is a very bad sign.

I could give you articles on why this is bad, or rant about it, but I'll try to be brief. Global's live for the lifetime of your program, so they are inherently inefficient. More importantly, Globals can be used anywhere. If you modify game_over in lots of places in your program it becomes hard to track down whether it was set to true at the right time, or if you need to check right now if its still false because it might have been set to true somewhere else.

File: CppTicTacToe/tictactoe.cpp [hunk 2]
```diff
@@ -13,16 +13,18 @@ using namespace std;
...
 // main global variables and initialization follow here
 
+// Tony Comment
 int round_counter = 1;
 char mark = 'X';
 bool game_over = false;
 int play_again_input = 0;
 int spot_to_fill = 0; // for empty spots, until filled by player and used by player_move() and attempt_to_fill_spot()
+// Tony Comment
 char board_spots[9] = {'1', '2', '3', '4', '5', '6', '7', '8', '9'};
```
Comment (ajscimone) [inline CppTicTacToe/tictactoe.cpp:23]: Instead of using an old school c array, you should use an actual std::array here. It will serve you much better.

File: CppTicTacToe/tictactoe.cpp [hunk 3]
```diff
@@ -13,16 +13,18 @@ using namespace std;
...
 
 
 // function prototypes follow here
-
+// Tony Comment
 void draw_board();
```
Comment (ajscimone) [inline CppTicTacToe/tictactoe.cpp:28]: Declaring all your function declarations here shows a lot about your code. Every function accepts no parameters and returns nothing. This makes things very hard to follow and reason about, even for you, the author.

Lets talk about the function declarations themselves. You dont actually need these. A function declaration is used as a place holder in c++. It basically says "im going to make this function later, i promise, just let me use it for now."

The compiler in C++ scans our code linearly. If it finds you try to use a function before its been declared, it will error. If you make a function declaration, and then make the function after you use it, that is allowed. It is, generally, bad practice to do so (unless you have a good reason).

File: CppTicTacToe/tictactoe.cpp [hunk 4]
```diff
@@ -65,7 +67,7 @@ void announce_three_in_a_row(){
 	// refactor the find_three_in_a_row() function.
 }
 
-
+// Tony Comment
 void round_limit_checker(){
```
Comment (ajscimone) [inline CppTicTacToe/tictactoe.cpp:71]: Now that we have said all that, lets talk about some things you do in your approach that we can work on. Lets take the round limit checker function. This function does a couple of things. It talks to the user, it checks if the game is over, and in increments the round. This is just too many things for this function to be responsible for. Instead, I want you to try and make functions that are responsible for an thing, and commincate back.

```
bool round_limit_checker(int round_counter){
    return round_counter > 9;
}
```
This is a much better function. This function takes the round counter object, and returns back true if we have reached the limit, or false if the game is still going.

We can use a function like this to check whether we have reached the round limit after each turn. This is the style and approach I would like you to try and take with all your functions. This might be hard, right now, but even being able to do some of them will be a huge improvement on the code base, and as we go, I will help you along the way. This function might not be the best to start with even, because it is currently deeply embedded in your code, but it was an easy on to demonstrate with.

File: CppTicTacToe/tictactoe.cpp [hunk 5]
```diff
@@ -93,6 +95,7 @@ void display_info(){
 	// I could mess up the game_logic flow.
 }
 
+// Tony Comment
 void find_three_in_a_row(){
```
Comment (ajscimone) [inline CppTicTacToe/tictactoe.cpp:99]: Above I talked about making functions that do a single thing for us, and that take in data, perform a calculation (in a manner of speaking) and communicate back their findings. I want you to start the process of turning your code into code that does that here with this function. Make the `find_three_in_a_row` take in the array, search for a win condition, and return true or false if it finds that a player has won the game or not.

File: CppTicTacToe/tictactoe.cpp (hunk 1 above)
Comment (ajscimone) [reply CppTicTacToe/tictactoe.cpp:17]: Eliminating your globals is a problem that will take work. Thankfully, it is rooted in dealing with the other aspects of your code that we can work on. You will see how we can eliminate globals as we reform your functions into things with responsibilities :)

**Review by CardioLeo — APPROVED** [review]

File: CppTicTacToe/tictactoe.cpp (hunk 1 above)
Comment (CardioLeo) [reply CppTicTacToe/tictactoe.cpp:17]: This makes sense about globals.

I guess I figured that in a smaller program like this, it didn't matter to just make most of them global scope. But it also makes sense to get closer to best practices even with small projects.

With a program like this, it feels like there is a pretty linear and logical way in which the globals are changed. It will be good practice to think about how not using globals can make the code more semantically focused, though.

File: CppTicTacToe/tictactoe.cpp (hunk 2 above)
Comment (CardioLeo) [reply CppTicTacToe/tictactoe.cpp:23]: I updated this just now and pushed it; but I'm not sure why I don't see the new commit on the review-responses branch...

Comment (CardioLeo) [reply CppTicTacToe/tictactoe.cpp:23]: Okay, got it now. The new commit shows up on review-responses now.
